Module update and testing 1.19

angelbarrera92 opened the pull request:
Hi team!

This PR contains a new update of the logging packages including a major upgrade of elasticsearch.

Includes:
- Add checks in the CI and fix policeman issues (super-linter).
- Add checks in the CI and fix license headers.
- Add 1.19 e2e testing pipeline.
- Add an automatic release pipeline.
- Update cerebro. From version `0.8.5` to `0.9.2`.
- Update fluentd. From version `1.10.2` to `1.11.2`.
- Update fluent-bit. From version `1.4.4` to `1.5.6`.
- Update elasticsearch. From version `6.8.8` to `7.9.1`
- Update kibana. From version `6.8.8` to `7.9.1`

The pipeline is passing right now: http://ci.sighup.io/sighupio/fury-kubernetes-logging/278/7/3

Thanks!

lnovara commented on docs/releases/v1.6.0.md:
```diff
@@ -0,0 +1,73 @@
+
+Once read, be sure to check and follow these steps:
+
+**VERY IMPORTANT NOTE:** *Be sure to have a recent backup of the elasticsearch indexes.*
```
Is there an easy way to backup the ES indexes that we can employ before doing the upgrade? Is a Velero volume snapshot enough?

angelbarrera92 replied:
What I can suggest is two backup in two different ways:
- 1. VolumeSnapshot with Velero could be a good idea
- 2. ElasticSearch snapshots using an S3 compatible service (minio, AWS, GCP...)

lnovara commented on katalog/elasticsearch-triple/triple.yml:
```diff
@@ -9,5 +13,5 @@ spec:
-            - name: discovery.zen.ping.unicast.hosts
-              value: "elasticsearch-0.elasticsearch,elasticsearch-1.elasticsearch,elasticsearch-2.elasticsearch"
+            - name: cluster.initial_master_nodes
+              value: "elasticsearch-0,elasticsearch-1,elasticsearch-2"
```
Does it make sense to use the DNS based discovery? This way, we could leverage the headless Kubernetes service populated by the StatefulSet.

https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-discovery-hosts-providers.html#built-in-hosts-providers

angelbarrera92 replied:
🤔 Don't know what's the difference with the current behavior

Reuenthal767 replied:
I may be wrong but I think that you cannot leverage the headless service when you are using `cluster.initial_master_nodes` (which is the first time you start the cluster, ever, clean volumes I believe).

You can use the headless service with `discovery.seed_hosts` but maybe we could call it `elasticsearch-headless` instead of just `elasticsearch` ^^

angelbarrera92 replied:
Currently, the `elasticsearch` service is of the type `headless`:

```yaml
clusterIP: None
```

This didn't change in this release. We are not deploying a non-headless service.

lnovara reviewed (approved):
LGTM :shipit: 

Left some non-blocking comments.

Reuenthal767 commented on katalog/elasticsearch-single/elasticsearch.yml:
```diff
@@ -90,7 +90,7 @@ spec:
+            - name: network.host
               value: "0.0.0.0"
             - name: ES_JAVA_OPTS
               value: -Xms2G -Xmx2G
```
I would add the following comment here: 
# Java heap must be 50% of total ram limits and min/max must be equal to avoid heap runtime resizing (a very costly process)

angelbarrera92 replied:
Currently, we are specifying to have the top 3Gi of ram. 
Do you think its a good idea to change the limit to 4Gi leaving the Xms and Xmx to 2Gi?

angelbarrera92 replied:
Or maybe I can specify less Xms Xmx (1.5) with the current limit of 3Gi

angelbarrera92 replied:
✅

Reuenthal767 commented on katalog/cerebro/README.md:
```diff
@@ -5,12 +5,12 @@ common tasks on ElasticSearch cluster via Cerebro's web interface.
 ## Requirements
 
-- Kubernetes >= `1.14.0`
+- Kubernetes >= `1.16.0`
```
Silly question: why do we need kubernetes >= 1.16.0 for Cerebro?

angelbarrera92 replied:
It is not a strict requirement. As we are testing all the packages from 1.16 up to 1.19, we can not ensure it will work in version < 1.16. This is why I set Kubernetes 1.16 as a requirement

Reuenthal767 commented:
I have some general questions:

- we are using a very basic readiness probe (which gave me a lot of problems with cluster nigthly shutdown) I think it would be valuable to use a more refined one (I can share some code).

- we are using a liveness probe, is it really needed with elasticsearch? If it's indeed the case I believe that the `initialDelaySeconds: 60` is not enough; I would at least double that, probably triple that to stay on the safe side.

- we are not using `podManagementPolicy: Parallel` but instead we are using the default ordered behaviour, why?

- In `fluentd-index-sighup-template.json` I would change the default replicas to 1, and add `"codec" : "best_compression"` in the settings. It decrease the size of logs by 15-25% and the cpu impact while it exists it's invisible to the human eye^^

angelbarrera92 commented:
> I have some general questions:
> 
>     * we are using a very basic readiness probe (which gave me a lot of problems with cluster nigthly shutdown) I think it would be valuable to use a more refined one (I can share some code).
> 

Absolutely, I tried the upstream readiness probe script: https://github.com/elastic/helm-charts/blob/master/elasticsearch/templates/statefulset.yaml#L225

But, deploying a single elasticsearch cluster didn't work as it remains in the yellow state forever.
Maybe we can patch the `triple.yml` to modify the `readinessProbe` adding the upstream script.

>     * we are using a liveness probe, is it really needed with elasticsearch? If it's indeed the case I believe that the `initialDelaySeconds: 60` is not enough; I would at least double that, probably triple that to stay on the safe side.
> 

Modified to 2x both liveness and readiness. ✅ 

>     * we are not using `podManagementPolicy: Parallel` but instead we are using the default ordered behaviour, why?

Modified! ✅ 

>     * In `fluentd-index-sighup-template.json` I would change the default replicas to 1, and add `"codec" : "best_compression"` in the settings. It decrease the size of logs by 15-25% and the cpu impact while it exists it's invisible to the human eye^^

✅

lnovara commented on katalog/elasticsearch-triple/triple.yml:
```diff
@@ -9,5 +13,66 @@ spec:
+                - sh
+                - -c
+                - |
+                  #!/usr/bin/env bash -e
```
🤯

angelbarrera92 replied:
Copy pasted from elasticsearch helm chart. Tested along with @Reuenthal767 ! Nice team-work

Reuenthal767 replied:
Yeah @lnovara that was my same reaction in the beginning, but what this readiness probe is doing is simply using a cheap trick (writing a file to /tmp) to differentiate between `cluster starting up` and `cluster already running` so that a suitable check can be performed.

I have used a similar readiness probe (just tailored for ES6) on 9 ES clusters for 2 weeks with very good results (especially regarding the nightly shutdown/restart of the underlying k8 cluster).

Also from what I've read in various github issues this readiness probe should help ensuring ES version upgrades are smooth.

Reuenthal767 commented on katalog/elasticsearch-triple/triple.yml:
```diff
@@ -9,5 +13,66 @@ spec:
+                    if [[ ${HTTP_CODE} == "200" ]]; then
+                      exit 0
+                    elif [[ ${HTTP_CODE} == "503" && "7" == "6" ]]; then
+                      exit 0
```
Sorry for my OCD, we can safely delete these lines as they are never used:
```
elif [[ ${HTTP_CODE} == "503" && "7" == "6" ]]; then
  exit 0
```

angelbarrera92 replied:
Nice catch up!

Reuenthal767 reviewed (approved):
LGTM!
